src/main/java/com/fasterxml/jackson/databind/deser/std/JsonNodeDeserializer.java

```diff
@@ -755,12 +755,11 @@ protected final JsonNode _fromFloat(JsonParser p, DeserializationContext ctxt,
             return _fromBigDecimal(ctxt, nodeFactory, p.getDecimalValue());
         }
         if (ctxt.isEnabled(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)) {
-            // 20-May-2016, tatu: As per [databind#1028], need to be careful
-            //   (note: JDK 1.8 would have `Double.isFinite()`)
-            if (p.isNaN()) {
-                return nodeFactory.numberNode(p.getDoubleValue());
+            try {
+                return _fromBigDecimal(ctxt, nodeFactory, p.getDecimalValue());
+            } catch (NumberFormatException nfe) {
+                // fall through - BigDecimal does not support values like NaN
```

Comment:
I don't think this makes sense -- this is an error condition and we should report an error shouldn't we?
(existing code was wrong already, I think)

Reply:
Actually, let me re-think -- should probably merge this as suggested but file a follow-up issue to consider making it fail as expected.

And possibly do that with configurable `JsonNodeFeature` setting.

Reply:
Created #4194 as follow-up